### DISCUSSION TITLE
Is there any possible way to snapshot a screen with modal sheet on the screen

### DISCUSSION THREAD

**Discussion by looneytsci** [pointfreeco/swift-snapshot-testing#956] [General]
I'm trying to make a snapshot for a view, which can open bottomSheet via .fullScreenCover and transparent background. Looks like It captures only the view layer without bottomSheet layer. Can I make the snapshot of the full view hierarchy? Or can I implement some custom snapshotting strategy with modal screen capturing?

**Comment by looneytsci** [comment]
PS this is for SwiftUI project

**Comment by tilucasoli** [comment]
I can make a snapshot of a Alert using this code below, propably it'll work to sheets @looneytsci 

```swift
extension Snapshotting where Value: UIViewController, Format == UIImage {    
    public static func windowedImage(interfaceStyle: UIUserInterfaceStyle = .unspecified,
                                     precision: Float = 0.98,
                                     perceptualPrecision: Float = 1) -> Snapshotting {
        return Snapshotting<UIImage, UIImage>.image(
            precision: precision,
            perceptualPrecision: perceptualPrecision,
            scale: nil
        ).asyncPullback { viewController in
            Async<UIImage> { callback in
                UIView.setAnimationsEnabled(false)
                let window = UIApplication.shared.windows.first!
                window.rootViewController = viewController
                window.windowScene?.traitOverrides.userInterfaceStyle = interfaceStyle
                DispatchQueue.main.async {
                    let image = UIGraphicsImageRenderer(bounds: window.bounds).image { _ in
                        window.drawHierarchy(in: window.bounds, afterScreenUpdates: true)
                    }
                    callback(image)
                    UIView.setAnimationsEnabled(true)
                }
            }
        }
    }
}

```

**Reply by OguzYuuksel** [reply]
Hello @tilucasoli, is this working from SPM target?

**Comment by LostStudent** [comment]
Im facing the same issue. The code above doesn't snapshot a sheet.

**Comment by cprince-foreflight** [comment]
So far, this isn't working for me. I was trying to test this alert:

```
struct TestAlert: View {
    @State var showAlert = true

    var body: some View {
        Text("Hello, World")
            .alert(isPresented: $showAlert) {
                Alert(title: Text("Alert Message"))
            }
    }
}
```

The test I have is:
```
    func testAlert() throws {
        assertSnapshot(of: TestAlert(), as: .image(drawHierarchyInKeyWindow: true, layout: .device(config: .iPhone8)))
    }
```

**Comment by cprince-foreflight** [comment]
> Cat you try
> 
> ```
>  let vc = UIHostingController(rootView: TestAlert())
>  let window = UIApplication.shared.windows.first!
> 
> let navigation =  UINavigationController(rootViewController: vc)
> navigation.modalPresentationStyle = .overFullScreen
> window.rootViewController?.present(navigation, animated: false)
> assertSnapshots(of: UIViewController(), as: [.windowedImage()])
> ```

Is that last line supposed to be: 
```
assertSnapshots(of: navigation, as: [.windowedImage()])

```
?

**Comment by daksh-dls** [comment]
Any updates on this?

**Comment by ralfebert** [comment]
I found the approach from @tilucasoli described [here](https://github.com/pointfreeco/swift-snapshot-testing/discussions/956#discussioncomment-12302692) works in general for capturing a SwiftUI .sheet(): `UIView.setAnimationsEnabled(false)` disables the animation and one can screenshot it by putting it on the actual UIWindow after doing a "main queue flush".

I suppose this is a bit trickier to get to work with Swift Testing. I had to remove `@MainActor` from the test method so that it runs on a bg thread and then add another `DispatchQueue.main.async` in the asyncPullback; but this is bad because the view creation in the test method should also happen on the main thread. Here is an example project for this approach anyway: https://github.com/ralfebert/SheetScreenshotExample/tree/mainasync

I also did a prototype where I use my own "put view on UIWindow and render an image" code. As the test is running on the main thread itself, I used the ugly workaround of running `RunLoop.main.run(until: Date.now.addingTimeInterval(0.1))` before capturing the screenshot to get the sheet on screen. Here is an example project for that approach: https://github.com/ralfebert/SheetScreenshotExample/tree/renderimage

I am wondering, is there a better way to do the "main queue flush" when the test method is `@MainActor`?

**Reply by ralfebert** [reply]
I guess the correct approach would be not to use `@MainActor` for the test method so we have more control / can act on the main queue from the outside. This would be good anyway to automatically parallelize comparing the images.

Based on the renderimage example, I thought I could just run two separate `@MainActor` tasks like this:

```swift
@Test func exampleScreenshot() async throws {

    let screen = await Task { @MainActor in
        // this is necessary to disable animations for sheets
        UIView.setAnimationsEnabled(false)

        let screen = AppWindowScreen(view: { ContentView() })
        screen.update()
        return screen
    }.value
            
    let image = await Task { @MainActor in
        let image = screen.renderImage()
        screen.tearDown()
        return image
    }.value
    
    assertSnapshot(of: image, as: .image)
}
```

I thought this alone should make sure that all main queue events that are triggered by putting the view on the UIWindow are processed before taking the screenshot.

But this only works sometimes, sometimes it fails with the sheet not visible.

I guess the sheet only becomes visible by "looking at it/when it actually renders".

To fix this, one can either render a "prep image" before the actual screenshot image, this made it very reliable in my experiments. Alternatively, beating on the UIWindow with a stick also seems to do the trick:

```swift
self.window.setNeedsLayout()
self.window.layoutIfNeeded()
self.window.setNeedsDisplay()
CATransaction.flush()
```

Work in progress, prototype is in https://github.com/ralfebert/SheetScreenshotExample/tree/renderimage-bg

**Reply by ralfebert** [reply]
This is a tricky one!

I resorted to make the test methods `async @MainActor` methods because that was more practical for my existing test suite and I can snapshot sheets reliably using the `async @MainActor` tests because I can actually "pause" the test execution and run the Main runloop for real by calling Task.yield(). This is the only thing that made the sheets appear / disappear reliably; if I use a non-async test method running on the main thread and use Runloop.main.run to flush the main queue, it works somewhat, but sometimes sheets don't disappear.

But with this method, the main queue runs freely "out-of-test-context", e.g. when using withDependencies, this will use the live dependencies because it cannot know that it's running in a test. Ouch.

🤯

**Reply by ralfebert** [reply]
For now I'll leave things as they are, my sketchy version with the `@Dependency` issue has to do until iOS 26 is released.

It works reliable enough for now in my use case (converting my app from a custom-built sheet to the system-sheet on iOS 26). The dependency injection issue seems to happen during cleanup, it doesn't influce how the views are presented. But it feels whacky and definitively needs more work.

If anyone is interested, I updated the prototype project with my current version of the approach: https://github.com/ralfebert/SheetScreenshotExample/tree/renderimage-bg

I want to find a proper solution for this eventually and contribute support for adding sheets to swift-snapshot-testing. If somebody deeply interested in Swift Testing / snapshot testing / pointfreeco's swift-dependencies / intricate details of SwiftUI/UIKit threading behavior wants to work on this together, let me know; it would be great to have a sparring partner to move this forward.
